Just like in Template-driven Forms, we can add our custom validators in Reactive Forms. And to make everything simple, let's take the same example as we did in Template-driven forms. 

So, we want to make sure the user cannot enter a username that is "admin".

In the Template-driven Forms, we had to create a new Directive, then we had to implement the "Validator" interface and then we had to implement the "validate" method inside it.

But, in Reactive forms, the process is a bit simpler.

We can add it in the component.ts file itself. 

We just need to add it as a normal function 

    forbiddenNames = ['admin', 'sudo']
    forbiddenName(control: FormControl): ValidationErrors {

        if (this.forbiddenNames.includes(control.value)){
        return {"forbiddenName": true, "notAllowed": control.value}
        }

        return null
    }

So, this is a normal function that takes the "Form Control" as the argument and returns "ValidationErrors" object which is a simple key-value pair where the "key" is a string and "value" can be any type.

This is something we returned in custom validator in the Template-driven forms as well.

And now, to use it on username field, we just do -

    username: new FormControl('', [Validators.required, this.forbiddenName]),


# ERROR DUE TO "this" KEYWROD

But now, there is an issue. The console will throw an error saying - 

    Cannot read properties of undefined (reading 'forbiddenNames')


So, for some reason, even though we have defined an array of strings "forbiddenNames" in our component, Angular says "this" is undefined since we are accessing "forbiddenNames" by doing "this.forbiddenNames".

And that's something that is related to Javascript or Typescript in general.

In our custom validator, we are using "this" keyword to access the "forbiddenNames" property. While this works as expected when we call this method inside the class itself, it will not work as expected when Angular calls it from outside of this class to check the validity of username field.

Because "this" always refers to what calls it. 

There are two ways to fix this issue.

1. Using Arrow Functions - We know that when we use Arrow Function syntax, the "this" keyword always represents the object that defined the arrow function. It means, no matter from where it is called, "this" will always represent the component.ts class.

    forbiddenName = (control: FormControl): ValidationErrors => {

        if (this.forbiddenNames.includes(control.value)){
        return {"forbiddenName": true, "notAllowed": control.value}
        }

        return null
    }

2. Using "bind" - If we don't want to use Arrow Function syntax, we have to use the "bind" method when we add this validator to the username form control - 
   
   username: new FormControl('', [Validators.required, this.forbiddenName.bind(this)])

When we use "bind" method, "this" will always refer to the place where it is defined, that is, the component.ts file.

Personally, I like to use the arrow function syntax. :D

# SHOWING ERROR MESSAGES BASED ON DIFFERENT VALIDATORS

Now that we have two different validators on username field, let us show two different error messages.

    <p
        class="text-danger"
        *ngIf="signupForm.get('userData.username').invalid && signupForm.get('userData.username').touched"
    >
        <!-- REQUIRED VALIDATION FAILS -->
        <span *ngIf="signupForm.get('userData.username').errors?.['required']"> Username cannot be empty!</span>
            
        <!-- FORBIDDEN NAME VALIDATION FAILS -->
        <span *ngIf="signupForm.get('userData.username').errors?.['forbiddenName']">
            Username {{signupForm.get('userData.username').errors?.['notAllowed']}} is not allowed!
        </span>
    </p>

You can see how we are using the "forbiddenName" and "notAllowed" keys that we returned when the forbiddenName validation failed.

